packages/react-devtools-shared/src/backend/renderer.js

```diff
@@ -522,13 +524,191 @@ export function attach(
+    } else {
+      messageID = freeErrorOrWarningMessageID++;
+      errorOrWarningToId.set(message, messageID);
+      idToErrorOrWarning.set(messageID, message);
+    }
+
+    const errorOrWarningsForFiber: RecordedErrorsOrWarnings = errorsOrWarnings.get(
+      fiberID,
+    ) || {
+      errors: [],
+      warnings: [],
+    };
+    if (type === 'warn') {
+      errorOrWarningsForFiber.warnings.push(messageID);
+    } else if (type === 'error') {
+      errorOrWarningsForFiber.errors.push(messageID);
+    }
+    errorsOrWarnings.set(fiberID, errorOrWarningsForFiber);
+  }
+
+  function clearErrorsAndWarnings() {
+    pendingErrorOrWarnings.splice(0);
+    clearTimeout(flushErrorOrWarningUpdatesTimoutID);
+    flushErrorOrWarningUpdatesTimoutID = null;
+
+    const updatedFiberIDs = new Set(errorsOrWarnings.keys());
+
+    freeErrorOrWarningMessageID = 0;
+    errorsOrWarnings.clear();
+    errorOrWarningToId.clear();
+    idToErrorOrWarning.clear();
+
+    emitErrorOrWarningUpdates(updatedFiberIDs);
+  }
+
+  function clearErrorsForFiberID(id: number) {
+    if (errorsOrWarnings.has(id)) {
+      const errorsOrWarningsForFiber: RecordedErrorsOrWarnings = (errorsOrWarnings.get(
+        id,
+      ): any);
+      errorsOrWarningsForFiber.errors.splice(0);
+      emitErrorOrWarningUpdates(new Set([id]));
+    }
+  }
+
+  function clearWarningsForFiberID(id: number) {
+    if (errorsOrWarnings.has(id)) {
+      const errorsOrWarningsForFiber: RecordedErrorsOrWarnings = (errorsOrWarnings.get(
+        id,
+      ): any);
+      errorsOrWarningsForFiber.warnings.splice(0);
+      emitErrorOrWarningUpdates(new Set([id]));
+    }
+  }
+
+  function onErrorOrWarning(
+    fiber: Fiber,
+    type: 'error' | 'warn',
+    args: Array<any>,
+  ): void {
+    // There are a few places that errors might be logged:
+    // 1. During render (either for initial mount or an update)
+    // 2. During commit effects (both active and passive)
+    // 3. During unmounts (or commit effect cleanup functions)
+    //
+    // Initial render presents a special challenge-
+    // since neither backend nor frontend Store know about a Fiber until it has mounted (and committed).
+    // In this case, we need to hold onto errors until the subsequent commit hook is called.
+    //
+    // Passive effects also present a special challenge-
+    // since the commit hook is called before passive effects, are run,
+    // meaning that we might not want to wait until the subsequent commit hook to notify the frontend.
+    //
+    // For this reason, warnings/errors are sent to the frontend periodically (on a timer).
+    // When processing errors, any Fiber that is not still registered is assumed to be unmounted.
+    pendingErrorOrWarnings.push({
+      fiber,
+      type,
+      // React pushes the component stack to the args later.
+      // These are not interesting to the view in React devtools where we already have access to the component stack.
+      args: args.slice(),
```

Comment:
This seems a _little_ fragile, since it relies on our hook being called _before_ the component stack gets appended. That's probably okay so long as we have clear comments in both places. (I'll add one.)

Reply:
Actually I think a better idea would be for us to auto-remove the component stack from warnings and errors. That would strip out even the ones React adds. :)

Reply:
I want that into the conole override which is also responsible for mutating the original args. We need to pass filtered args anyway if they already contain the component stacks so we can copy them directly at that point. Then it's a bit more obvious if the copy happens before or after mutation.

Reply:
Check out 674e287 for how I thought we might deal with both cases (React stacks and appended stacks).